### DISCUSSION TITLE
Windows 10 Start Menu - Pin Weather / Clock Live Tiles?

### DISCUSSION THREAD

**Discussion by doterik** [valinet/ExplorerPatcher#1037] [Q&A] [answered]
I now see that the Windows 10 start menu has got a really nice look, a brilliant job I must say. 👍🏼

My problem is that I can't pin new places from the Weather app. Removing some went (of course) well but now I want to be able to add some new places. (Works on Windows 11 start menu, without live tiles, that is.)

Have looked through the forum but haven't found any answer. Is there any trick to make this work?

In fact, I don't know where the pinned apps are specified, have to Bing I suppose, but hoped this would work anyway.

Any tips on how to pin live tiles from the Weather app on Windows 10 start menu?

**Comment by surya-sk** [comment]
I'd like to know this too, for pinning different countries from the Clock app

**Comment by valinet** [comment, marked as answer]
tl;dr Skip to last bullet point for a workaround that works for now.

I explained in a couple of places already. From what I have learnt:

-> The Windows 11 Start menu uses a slightly different tiles storage back-end than the Windows 10 Start menu.
-> Pinning to Start is intermediated by `StartTileData.dll`, as well as other things probably. 
-> In Windows 11, the system `StartTileData.dll` in `C:\Windows\System32` ties to the newer backend that pins stuff to the Windows 11 menu.
-> This is the reason why pinning doesn't work from Windows Search, Clock, Weather and what not out of the box, as those simply request the `StartTileData` component (and some "thing" from it - this implements a ton of stuff and is registered for multiple Windows Runtime classes and COM interfaces) and the OS simply loads the version in `C:\Windows\System32`.
-> For the Windows 10 Start menu, ExplorerPatcher makes `StartMenuExperienceHost.exe` use a version of `StartTileData.dll` from OS build 22000.51 which still has the code for pinning stuff to the Windows 10 menu.

Technically, the newer `StartTileData.dll` still has code for pinning to the old Start menu as well. What has changed is that older versions of the library used the `Start_ShowClassicMode` registry entry to choose which tile backend to use, while newer versions mostly prefer to determine this by querying the Win32 `GetProductInfo` function. They will enable the usage of the tiles backend for the Windows 10 menu only if `GetProductInfo` returns `0x77` (`119`), which means "Windows PPI edition" aka "Windows Team" aka "Windows for Surface Hub". This is mostly inline with the actual reality: if you use some custom tool to make a Windows 11 Team ISO, you will see that the UI doesn't resemble Windows 11 at all, but rather it uses the legacy UI from Windows 10 still.

-> To pin stuff from File Explorer to the Windows 10 menu, ExplorerPatcher offers a shortcut in "Properties - Start menu - Pin tiles to Windows 10 Start menu from File Explorer". This opens a new File Explorer window you can use to locate whatever you'd like to pin there; right click and choose "Pin to Start". if it doesn't work the first time, do it once more, it eventually works. This window is special in that it is hosted in a separate instance of `explorer.exe` where ExplorerPatcher injects the older version of `StartTileData.dll`. The reason why I don't inject all instances is because some interfaces in the DLL have changed a bit; using the older DLL leads to weird things, like the taskbar not showing UWP apps in Insider Preview builds, for example. It's complicated...

-> Of course, from a technical standpoint, the proper solution is to patch the newer DLL to have it use the old storage backend when the Windows 10 Start menu is enabled. This is problematic for a number of reasons. As `StartTileData.dll` lives in `C:\Windows\System32`, classic approaches like patching by exploiting the library search order, as EP does in plenty of instances, is out of the question. Static patching as well, for obvious reasons (like, Windows Update fucking up the patch every 2 weeks).

-> Something that has real chances of actually working though is to determine which component that `StartTileData.dll` implements is exactly responsible for pinning things to Start. Then, override only the data store for that component, direct it to a proxy DLL of ours, maybe redirecting non-interesting interfaces to the original DLL and just the one we are interested in (the interface that does the pinning thing) to the old DLL that works with the old Start menu. Alternatively and probably more simply, since our proxy DLL would load when the component responsible for pinning stuff is requested, we can at that point chain load the original library, patch out `GetProductInfo`, and forward all calls from our DLL to the original DLL. In theory is doable, in practice I think it's doable as well, it just takes, as usual, time and effort. Something along these lines should work.

-> To determine which component is responsible for that, one has to query the registry. Of interest are the following paths, in particular:

```
HKEY_LOCAL_MACHINE\SOFTWARE\Classes\CLSID
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\WindowsRuntime\ActivatableClassId
```

This represent the registry for COM and Windows Runtime objects, respectively. The system uses entries in these keys to determine which library hosts which component. Overrides have to be made here, redirecting things from the system's `StartTileData.dll` to the older one. Do this in a VM, let's say, change the first thing you find, kill `ApplicationFrameHost.exe`, launch Weather, try to pin to Start. If it works, then the last changed thing is probably what we are interested in. If it does not, find next thing mentioning `StartTileData.dll`, repeat steps above. Something along these lines.

-> As a temporary workaround that I have tested and works, you can do this:

* Make sure to have the Windows 10 Start menu enabled in EP Properties.
* In `C:\Windows\System32`, take ownership of `StartTileData.dll` and rename it to `StartTileData_original.dll`.
* Copy `StartTileDataLegacy.dll` from `C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy` as `StartTileData.dll` in `C:\Windows\System32`. `StartTileDataLegacy.dll` is untouched `StartTileData.dll` from OS build 22000.51.
* Close UWP apps and kill `ApplicationFrameHost.exe` (which is responsible for hosting UWP apps, like Weather, Microsoft News etc) and reload the app that you want to pin things from (for example, Weather).
* Pin the thing. It will work. You can check that in the Windows 10 Start menu.
* When you're done, rename `StartTileData.dll` to `StartTileDataLegacy.dll` in `C:\Windows\System32` and restore `StartTileData_original.dll` as `StartTileData.dll`. Close UWP apps and kill `ApplicationFrameHost.exe` or reload the app that you were using to have it load the system-shipped library.

-> Why I don't recommend leaving the old library in place? Besides Windows Update which will overwrite it anyway on the next update, we don't really know if it works well. With how opaque everything is, in some scenario it might fuck something up that they changed and the old library doesn't have a clue about. For these concerns, it's best to minimize the time using this exotic config, in my opinion. Ofc, it's your computer and your knowledge, and progress is done via experimentation, so take this as a reminder of what *might* happen. It *might* as well be totally fine, for now at least, to run daily like that, I simply don't have a definite answer here.

-> Things will pin. Will they work? Live tiles I mean. Idk. Stuff seems pretty broken still. Or maybe when they update the apps to make them look like Windows 11, they take out live tiles support, since the new Start features just useless icons...? Idk really, I simply haven't coded these things, I just look at them from the outside, same as you do. The live tile for Weather seems to work, so there's still some support left. Idk. 

-> Tbh, personally, the most I have used these live tiles was for weather. Which now I have replaced with the weather widget. Other than that, I took the time to enable this Windows 10 Start menu because it looked doable from what I have researched previously and liked its layout more (app list and an area to pin some apps on the same page makes too much sense). I don't like locking myself to using proprietary things that only one party controls, and Windows 10 Start and live tiles are and were just that. Don't get locked into using proprietary ecosystems. Sure, use them, but don't become their prisoners. Don't let momentary convenience win over long time actual freedom. Why? You have the example here: when they feel like it, the owner walks away, brakes everything and you're left scratching your head, they don't care. The black box is gigantic, takes ages to fully grasp. My advice, at its core, is the same as always - if you relied on these Windows 10 specific things so much, best is to stay on 10, at least as long as those work there (the apps will eventually become outdated and the cloud backends they use will be shutdown and so will be unable to feed them data). But really, I don't advise that - instead, use some browser web page with widgets, idk, there should be plenty of things on GitHub, something that's open, and add data sources there. Browsers will always work, and open things are in many cases easier to follow, understand, maintain, update and fix, by you or the community, as opposed to these black boxes that you have no control on. I bet you in 10 years they will throw the Windows 11 shell to the trashcan again, if they still care for the Windows as an OS by then.

**Reply by doterik** [reply]
Amazing, it really worked ... 😀

Thanks for your answer, which more than well shed light on the problem.

Not as smooth as I had hoped for, but I can hardly blame you for that. 😉
With a bat-file relatively smoothly though. Thanks! 👍🏼

/*
Tbh, I also only used live tiles for weather myself, which I missed with Windows 11.

Clock, as someone mentioned, could be another example, but I'm satisfied with the two extra zones I have on the clock in the taskbar.

About legacy code, one of the reasons why I haven't used the Windows 10 start menu so far but saw that it has now become quite nice and modern so I will try it again for a while. (But, without weather tiles, it's pretty uninteresting to me... and I can also live without them...)
*/

Thanks for a masterful answer and keep the great work with the app going. 👏🏼

**Reply by valinet** [reply]
That’s great to hear. Thanks for the feedback as well. Hopefully more research happens in this area and things may improve in the future.

Btw, if you don’t mind, you can post the batch file here, maybe it will help others automate this occasional task for their workflows.

Thank you

**Reply by doterik** [reply]
It works for Me ... on My machine ... no other guarantees ...
```
@echo off & cd/d c:\Windows\System32

if not exist .\StartTileData.dll.0LD (
  .\takeown.exe /f .\StartTileData.dll /a
  copy c:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy\StartTileDataLegacy.dll StartTileData.dll.0LD
)

ren .\StartTileData.dll StartTileData.dll.0RG
ren .\StartTileData.dll.0LD StartTileData.dll

.\taskkill.exe /f /im ApplicationFrameHost.exe

echo ** Time to do your work...
echo.
echo ** Press enter when done!
echo.
pause

ren .\StartTileData.dll StartTileData.dll.0LD
ren .\StartTileData.dll.0RG StartTileData.dll

.\taskkill.exe /f /im ApplicationFrameHost.exe
```
###### EDIT: I wasn't prepared to publish this, now I have done some minor changes to the text so that others may understand, not only me...
